tests/bluetooth/bsim_bt/bsim_test_l2cap/src/main_l2cap_ecred.c

```diff
@@ -0,0 +1,618 @@
+		for (int k = 0; k < L2CAP_CHANNELS; k++) {
+			channels[k].iteration = i;
+			/* Assign the right msg to the right channel */
+			channels[k].bytes_to_send = (k == LONG_MSG_CHAN_IDX) ? LONG_MSG : SHORT_MSG;
+			int err = k_work_submit_to_queue(&channels[k].work_queue,
+									&channels[k].work);
+
+			if (err < 0) {
+				FAIL("Failed to submit work to the queque, error: %d", err);
+			}
+		}
+
+		/* Wait until messages on all of the channels has been sent */
+		for (int l = 0; l < L2CAP_CHANNELS; l++) {
+			k_sem_take(&sent_sem, K_FOREVER);
+		}
+	}
+}
+
+static int change_mtu_on_channels(int num_channels, int new_mtu)
+{
+	int err;
+	struct bt_l2cap_chan *reconf_channels[ECRED_CHAN_MAX] = { NULL };
+
+	for (int i = 0; i < num_channels; i++) {
+		reconf_channels[i] = &(&channels[i])->le.chan;
+	}
+
+	err = bt_l2cap_ecred_chan_reconfigure(reconf_channels, new_mtu);
+
+	return err;
+}
+
+static void test_peripheral_main(void)
+{
+	int err;
+
+	device_sync_init(PERIPHERAL_ID);
+	LOG_DBG("*L2CAP ECRED Peripheral started*");
+	init_workqs();
+	err = bt_enable(NULL);
+
+	if (err) {
+		FAIL("Can't enable Bluetooth (err %d)", err);
+		return;
+	}
+
+	LOG_DBG("Peripheral Bluetooth initialized.");
+	LOG_DBG("Connectable advertising...");
+	err = bt_le_adv_start(BT_LE_ADV_CONN_NAME, ad, ARRAY_SIZE(ad), NULL, 0);
+
+	if (err) {
+		FAIL("Advertising failed to start (err %d)", err);
+		return;
+	}
+
+	LOG_DBG("Advertising started.");
+	LOG_DBG("Peripheral waiting for connection...");
+	WAIT_FOR_FLAG_SET(is_connected);
+	LOG_DBG("Peripheral Connected.");
+	register_l2cap_server();
+	connect_num_channels(L2CAP_CHANNELS);
+	k_sem_take(&all_chan_conn_sem, K_FOREVER);
+
+	/* Disconnect and reconnect channels *****************************************************/
+	LOG_DBG("#####################Disconnect and reconnect channels");
+	disconnect_all_channels();
+	k_sem_take(&all_chan_disconn_sem, K_FOREVER);
+
+	connect_num_channels(L2CAP_CHANNELS);
+	k_sem_take(&all_chan_conn_sem, K_FOREVER);
+
+	LOG_DBG("Send sync after reconnection");
+	device_sync_send();
+
+	/* Send bytes on both channels and expect ch 1 to receive all of them before ch 0 *********/
+	LOG_DBG("###################Send bytes on both channels concurrently");
+	send_sdu_concurrently();
+
+	/* Change mtu size on all connected channels *********************************************/
+	LOG_DBG("###################Change MTU of the channels");
+	err = change_mtu_on_channels(L2CAP_CHANNELS, CONFIG_BT_L2CAP_TX_MTU + 10);
+
+	if (err) {
+		FAIL("MTU change failed (err %d)\n", err);
+	} else {
+		PASS("MTU change Passed\n");
+	}
+
+	/* Read from both devices (Central and Peripheral) at the same time **********************/
+	LOG_DBG("##############Read from both devices (Central and Peripheral) at the same time");
+	LOG_DBG("Wait for sync before sending the msg");
+	device_sync_wait();
+	LOG_DBG("Received sync");
+	send_sdu(0, 1, 10);
+
+	disconnect_all_channels();
```

Comment:
Don't we risk disconnecting the channels too soon?

Reply:
dosen't seem so. The test is quite stable. Never failed after many trials

Reply:
But would it make sense to wait for the `sent` callback, before we disconnect? (I think so), otherwise we risk introducing a race condition here I think, that may cause problems in the future.